staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/generic/controller.go

```diff
@@ -0,0 +1,255 @@
+				if len(oldMeta.GetResourceVersion()) == 0 {
+					klog.Warningf("%v throwing out update with empty RV. this is likely to happen if a test did not supply a resource version on an updated object", c.options.Name)
+				}
+				return
+			}
+
+			enqueue(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// Enqueue
+			enqueue(obj)
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	//!TODO: check if informer is even valid?
+	// e.g. if crd isnt even installed yet this just waits forever here
+	if !cache.WaitForNamedCacheSync(c.options.Name, ctx.Done(), c.informer.HasSynced) {
+		// ctx cancelled during cache sync. return early
+		err := ctx.Err()
+		if err == nil {
+			// if context wasnt cancelled then the sync failed for another reason
+			err = errors.New("cache sync failed")
+		}
+		return err
+	}
+
+	waitGroup := sync.WaitGroup{}
+
+	for i := uint(0); i < c.options.Workers; i++ {
+		waitGroup.Add(1)
+		go func() {
+			wait.Until(c.runWorker, time.Second, ctx.Done())
+			waitGroup.Done()
+		}()
+	}
+
+	klog.Infof("Started %v workers for %v", c.options.Workers, c.options.Name)
+
+	// Wait for context cancel.
+	<-ctx.Done()
+
+	// Gracefully shutdown workqueue. Finish processing any enqueued items.
+	//!TODO: May want to put deadline on this and forcefully shutdown?
+	c.queue.ShutDownWithDrain()
+
+	// Workqueue shutdown signals for workers to stop. Wait for all workers to
+	// clean up
+	waitGroup.Wait()
+
+	// Remove event handler and ignore error (only raised for improper usage)
+	c.informer.RemoveEventHandler(registration)
+
+	// Only way for workers to ever stop is for caller to cancel the context
+	return ctx.Err()
+}
+
+func (c *controller[T]) HasSynced() bool {
+	return c.informer.HasSynced()
+}
+
+func (c *controller[T]) runWorker() {
+	for {
+		obj, shutdown := c.queue.Get()
+		if shutdown {
+			return
+		}
+
+		// We wrap this block in a func so we can defer c.workqueue.Done.
+		err := func(obj interface{}) error {
+			// We call Done here so the workqueue knows we have finished
+			// processing this item. We also must remember to call Forget if we
+			// do not want this work item being re-queued. For example, we do
+			// not call Forget if a transient error occurs, instead the item is
+			// put back on the workqueue and attempted again after a back-off
+			// period.
+			defer c.queue.Done(obj)
+			var key string
+			var ok bool
+			// We expect strings to come off the workqueue. These are of the
+			// form namespace/name. We do this as the delayed nature of the
+			// workqueue means the items in the informer cache may actually be
+			// more up to date that when the item was initially put onto the
+			// workqueue.
+			if key, ok = obj.(string); !ok {
+				// As the item in the workqueue is actually invalid, we call
+				// Forget here else we'd go into a loop of attempting to
+				// process a work item that is invalid.
+				c.queue.Forget(obj)
```

Comment:
Not sure this will help -- how did an incorrectly formatted key get here in the first place? I assume if it happens once it'll keep happening. Just Done() (and no AddRateLimited) is sufficient to get it out of the queue.

Reply:
I agree with this reasoning - that it makes no sense to reset rate limiter for an object with an invalid key.

Similarly to the other comment, this quirk is also present in the sample controller.